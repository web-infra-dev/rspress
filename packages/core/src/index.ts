// TODO: do not expose remarkPluginNormalizeLink as publicAPI
export { dev, build, serve, remarkPluginNormalizeLink } from './node';
export * from '@rspress/shared';
export { mergeDocConfig } from '@rspress/shared/node-utils';
export type { RouteService } from './node/route/RouteService';
