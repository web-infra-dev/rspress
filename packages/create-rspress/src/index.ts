#!/usr/bin/env node
import path from 'node:path';
import { cac } from 'cac';
import prompts from 'prompts';
import fs from 'fs-extra';
import {
  copyFolder,
  initSitePrompts,
  formatTargetDir,
  getPkgManager,
  renderFile,
  cancelPrompt,
} from './utils';
import type { CustomPromptObject } from './types';

const cli = cac('create-rspress').help();

cli.command('', 'Create a new rspress site').action(async () => {
  const siteOptions: CustomPromptObject[] = [
    {
      name: 'siteTitle',
      type: 'text',
      message: 'Site Title',
      initial: 'Rspress',
      value: '',
    },
    {
      name: 'siteDesc',
      type: 'text',
      message: 'Site Description',
      initial: 'Rspack-based Static Site Generator',
      value: '',
    },
  ];

  const defaultProjectName = 'rspress-site';
  let targetDir = defaultProjectName;
  const promptProjectDir = async () =>
    await prompts(
      [
        {
          type: 'text',
          name: 'projectDir',
          initial: defaultProjectName,
          message: 'Project folder',
          onState: state => {
            targetDir = formatTargetDir(state.value) || defaultProjectName;
          },
        },
      ],
      { onCancel: cancelPrompt },
    );

  await promptProjectDir();
  let root = path.resolve(process.cwd(), targetDir);
  while (fs.existsSync(root)) {
    console.log(
      `${targetDir} is not empty, please choose another project name`,
    );
    await promptProjectDir();
    root = path.resolve(process.cwd(), targetDir);
  }

  await initSitePrompts(siteOptions);

  await fs.mkdir(root, { recursive: true });
  const srcFolder = path.resolve(__dirname, '../template');
  await copyFolder(srcFolder, targetDir);

  const filesToInterpret = ['docs/index.md', 'rspress.config.ts'];
  const siteData = siteOptions.reduce((prev: Record<string, any>, cur) => {
    prev[cur.name as string] = cur.value
      .replace(/\\/g, '\\\\')
      .replace(/'/g, "\\'");
    return prev;
  }, {});
  for (const file of filesToInterpret) {
    await renderFile(file, targetDir, targetDir, siteData);
  }

  const pkgManager = getPkgManager();
  console.log('\nDone. Now run:\n');
  console.log(`cd ${targetDir}\n`);
  console.log(`${pkgManager} install\n`);
  console.log(`${pkgManager} run dev\n`);
});

cli.parse();
