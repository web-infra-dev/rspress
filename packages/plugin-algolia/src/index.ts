import type { RspressPlugin } from '@rspress/shared';

interface Options {
  /**
   * Algolia meta tag for verification
   * - <meta name="algolia-site-verification" content="YOUR_VERIFICATION_CONTENT" />
   */
  verificationContent?: string;
}

export function pluginAlgolia(options: Options = {}): RspressPlugin {
  const { verificationContent } = options;
  return {
    name: '@rspress/plugin-algolia',
    config(config) {
      config.search = false;
      return config;
    },
    builderConfig: {
      html: {
        meta: verificationContent
          ? {
              'algolia-site-verification': {
                name: 'algolia-site-verification',
                content: verificationContent,
              },
            }
          : {},
      },
    },
  };
}
