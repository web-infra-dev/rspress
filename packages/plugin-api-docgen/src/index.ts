import path from 'path';
import type { RspressPlugin } from '@rspress/shared';
import fs from '@modern-js/utils/fs-extra';
import type {
  PluginOptions,
  SupportLanguages,
  ExtendedPageData,
} from './types';
import { docgen } from './docgen';
import { apiDocMap } from './constants';

/**
 * The plugin is used to generate api doc for files.
 */
export function pluginApiDocgen(options?: PluginOptions): RspressPlugin {
  const {
    entries = {},
    apiParseTool = 'react-docgen-typescript',
    appDir = process.cwd(),
    parseToolOptions = {},
  } = options || {};
  return {
    name: '@modern-js/doc-plugin-api-docgen',
    config(config) {
      config.markdown = config.markdown || {};
      config.markdown.mdxRs = false;
      return config;
    },
    async beforeBuild(config, isProd) {
      // only support zh and en
      const languages = (
        config.themeConfig?.locales?.map(locale => locale.lang) ||
        config.locales?.map(locale => locale.lang) || [config.lang]
      ).filter(lang => lang === 'zh' || lang === 'en') as SupportLanguages[];
      await docgen({
        entries,
        apiParseTool,
        languages,
        appDir,
        parseToolOptions,
        isProd,
      });
    },
    async modifySearchIndexData(pages) {
      // Update the search index of module doc which includes `<API moduleName="foo" />` and `<API moduleName="foo" ></API>
      const apiCompRegExp = /(<API\s+moduleName=['"](\S+)['"]\s*(.*)?\/>)|(<API\s+moduleName=['"](\S+)['"]\s*(.*)?>(.*)?<\/API>)/;
      await Promise.all(
        pages.map(async page => {
          const { _filepath, lang } = page;
          let content = await fs.readFile(_filepath, 'utf-8');
          let matchResult = new RegExp(apiCompRegExp).exec(content);
          if (!matchResult) {
            return;
          }
          while (matchResult !== null) {
            const [matchContent, moduleName] = matchResult;
            const apiDoc =
              apiDocMap[moduleName] || apiDocMap[`${moduleName}-${lang}`];
            content = content.replace(matchContent, apiDoc);
            matchResult = new RegExp(apiCompRegExp).exec(content);
          }
          page.content = content;
        }),
      );
    },
    extendPageData(pageData) {
      (pageData as ExtendedPageData).apiDocMap = { ...apiDocMap };
    },
    markdown: {
      globalComponents: [
        path.join(__dirname, '..', 'static', 'global-components', 'API.tsx'),
      ],
    },
  };
}

export type { PluginOptions };
