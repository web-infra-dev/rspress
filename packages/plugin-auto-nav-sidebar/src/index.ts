import path from 'node:path';
import { type RspressPlugin, addTrailingSlash } from '@rspress/shared';
import { logger } from '@rspress/shared/logger';
import { combineWalkResult } from './utils';
import { walk } from './walk';

// Scan all the directories and files in the work directory(such as `docs`), and then generate the nav and sidebar configuration according to the directory structure.
// We will do as follows:
// 1. scan the directory structure, and extract all the `_meta.json` files.
// The `_meta.json` files will have two types:
// - For the `_meta.json` directly in the work directory, it will be used as the nav configuration.
// - For the `_meta.json` in the subdirectory, it will be used as the sidebar configuration.
// First, for the `_meta.json` directly in the work directory, the json content will have the following structure(see `NavMeta`):
// ```ts
// export type NavMeta = NavItem[];
// export type NavItem = NavItemWithLink | NavItemWithChildren;

// export type NavItemWithLink = {
//   text: string;
//   link: string;
//   activeMatch?: string;
//   position?: 'left' | 'right';
// };

// export type NavItemChildren = {
//   text: string;
//   items: NavItemWithLink[];
// };
// ```
// The `NavItemWithLink` will be used as the nav item, and the `NavItemChildren` will be used as the nav item with children.
// Second, for the `_meta.json` in the subdirectory, the json content will have the following structure(see `SideMeta`):
// ```ts
// export type SideMetaItem =
//   | string
//   | {
//       type: 'file' | 'directory';
//       name: string;
//       // Use the h1 title as the sidebar title by default
//       label?: string;
//       collapsible?: boolean;
//       collapsed?: boolean;
//     };

// export type SideMeta = SideMetaItem[];
// ```
// The `SideMetaItem` can be the following types:
// - string: the file name, such as `home.md`, can also remove the extension name, such as `home`.
// - object: you can specify the `type` and `name` of the file or directory, and the `label` of the sidebar item, and whether the sidebar item is `collapsible` and `collapsed`.

// 2. generate the nav and sidebar configuration according to the `_meta.json` files.

// As you can see, every `_meta.json` file will have a corresponding file or directory, so we can get the link info from the file or directory(relative to the work directory).So we can generate the nav and sidebar config based on link info and _meta.json content.

// For nav config, we don't need to do anything, just use the `_meta.json` content as the nav config.
// For sidebar config, we need to transform every item of the array config:
// - string: we transform the string to a object with `text` and `link` property.Such as:
// ```ts
// {
//   text: 'home',
//   link: '/home'
// }
// ```
// - object: in this case, we will take account of the `type` property:
//   - file: we will transform the object to a object with `text` and `link` property.Such as:
//   ```ts
//   {
//     text: 'home',
//     link: '/home'
//   }
//   ```
//   - directory: we will transform the object to a object with `text` and `items` property.

// There is the following file structure:
// ```
// docs
// ├── guide
// │   ├── home.md
// │   ├── advanced
// │   │    └── plugin.md
// │   └── _meta.json
// └── _meta.json
// ```
// The `_meta.json` in the work directory has the following content:
// ```json
// [
//   {
//     "text": "guide",
//     "link": "/guide"
//   }
// ]
// ```
// The `_meta.json` in the `guide` directory has the following content:
// ```json
// [
//   'home',
//   {
//     "type": "directory",
//     "name": "advanced",
//     "label": "advanced",
//     "collapsible": true,
//     "collapsed": false
//   }
// ]
// ```
// The `home.md` has the following content:
// ```md
// # home
// ```
// The `_meta.json` in the `advanced` directory has the following content:
// ```json
// [
//   {
//     "type": "file",
//     "name": "plugin",
//     "label": "plugin",
//     "collapsible": true,
//     "collapsed": false
//   }
// ]
// ```
// The `plugin.md` has the following content:
// ```md
// # plugin
// ```

// The generated nav config will be:
// ```json
// [
//   {
//     "text": "guide",
//     "link": "/guide"
//   }
// ]
// ```
// The generated sidebar config will be:
// ```json
// [
//   {
//     "text": "home",
//     "link": "/guide/home"
//   },
//   {
//     "text": "advanced",
//     "items": [
//       {
//         "text": "plugin",
//         "link": "/guide/advanced/plugin"
//       }
//     ]
//   }
// ]
// ```

function processLocales(
  langs: string[],
  versions: string[],
  root: string,
  defaultLang: string,
  defaultVersion: string,
) {
  return Promise.all(
    langs.map(async lang => {
      const walks = versions.length
        ? await Promise.all(
            versions.map(version => {
              const routePrefix = addTrailingSlash(
                `${version === defaultVersion ? '' : `/${version}`}${
                  lang === defaultLang ? '' : `/${lang}`
                }`,
              );
              return walk(path.join(root, version, lang), routePrefix, root);
            }),
          )
        : [
            await walk(
              path.join(root, lang),
              addTrailingSlash(lang === defaultLang ? '' : `/${lang}`),
              root,
            ),
          ];
      return combineWalkResult(walks, versions);
    }),
  );
}

export function pluginAutoNavSidebar(): RspressPlugin {
  return {
    name: 'auto-nav-sidebar',
    async config(config) {
      config.themeConfig = config.themeConfig || {};
      config.themeConfig.locales =
        config.themeConfig.locales || config.locales || [];
      const langs = config.themeConfig.locales.map(locale => locale.lang);
      const hasLocales = langs.length > 0;
      const hasLang = Boolean(config.lang);
      const versions = config.multiVersion?.versions || [];
      const defaultLang = config.lang || '';
      const { default: defaultVersion = '' } = config.multiVersion || {};
      if (hasLocales) {
        const metaInfo = await processLocales(
          langs,
          versions,
          config.root!,
          defaultLang,
          defaultVersion,
        );
        config.themeConfig.locales = config.themeConfig.locales.map(
          (item, index) => ({
            ...item,
            ...metaInfo[index],
          }),
        );
      } else {
        if (hasLang) {
          logger.error(
            '`lang` is configured but `locales` not, ' +
              'thus `@rspress/plugin-auto-nav-sidebar` can not auto generate ' +
              'navbar level config correctly!\n' +
              'please check your config file',
          );
          return config;
        }
        const walks = versions.length
          ? await Promise.all(
              versions.map(version => {
                const routePrefix = addTrailingSlash(
                  version === defaultVersion ? '' : `/${version}`,
                );
                return walk(
                  path.join(config.root!, version),
                  routePrefix,
                  config.root!,
                );
              }),
            )
          : [await walk(config.root!, '/', config.root!)];

        const combined = combineWalkResult(walks, versions);

        config.themeConfig = { ...config.themeConfig, ...combined };
      }

      return config;
    },
  };
}
