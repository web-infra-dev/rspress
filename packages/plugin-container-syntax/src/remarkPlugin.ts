/**
 * ðŸš€ This plugin is used to support container directive in unified.
 * Taking into account the compatibility of the VuePress/Docusaurus container directive, current remark plugin in unified ecosystem only supports the following syntax:
 * ::: tip {title="foo"}
 * This is a tip
 * :::
 * But the following syntax is not supported:
 * ::: tip foo
 * This is a tip
 * :::
 * In fact, the syntax is usually used in SSG Frameworks, such as VuePress/Docusaurus.
 * So the plugin is used to solve the problem and support both syntaxes in above cases.
 */

import type { Plugin } from 'unified';
import {
  Root,
  Content,
  Parent,
  BlockContent,
  PhrasingContent,
  Paragraph,
  Literal,
} from 'mdast';

export const DIRECTIVE_TYPES = [
  'tip',
  'note',
  'warning',
  'caution',
  'danger',
  'info',
  'details',
] as const;
export const REGEX_BEGIN = /^\s*:::\s*(\w+)\s*(.*)?/;
export const REGEX_END = /\s*:::$/;
export const TITLE_REGEX_IN_MD = /{\s*title=["']?(.+)}\s*/;
export const TITLE_REGEX_IN_MDX = /\s*title=["']?(.+)\s*/;

export type DirectiveType = (typeof DIRECTIVE_TYPES)[number];

const trimTailingQuote = (str: string) => str.replace(/['"]$/g, '');

const parseTitle = (rawTitle = '', isMDX = false) => {
  const matched = rawTitle?.match(
    isMDX ? TITLE_REGEX_IN_MDX : TITLE_REGEX_IN_MD,
  );
  return trimTailingQuote(matched?.[1] || rawTitle);
};

/**
 * Construct the DOM structure of the container directive.
 * For example:
 *
 * ::: tip {title="foo"}
 * This is a tip
 * :::
 *
 * will be transformed to:
 *
 * <div class="rspress-directive tip">
 *   <div class="rspress-directive-title">TIP</div>
 *   <div class="rspress-directive-content">
 *     <p>This is a tip</p>
 *   </div>
 * </div>
 *
 */
const createContainer = (
  type: DirectiveType | string,
  title: string | undefined,
  children: BlockContent[],
): Parent => {
  const isDetails = type === 'details';

  const rootHName = isDetails ? 'details' : 'div';
  const titleHName = isDetails ? 'summary' : 'div';

  return {
    type: 'containerDirective',
    data: {
      hName: rootHName,
      hProperties: {
        class: `rspress-directive ${type}`,
      },
    },
    children: [
      {
        type: 'paragraph',
        data: {
          hName: titleHName,
          hProperties: {
            class: 'rspress-directive-title',
          },
        },
        children: [{ type: 'text', value: title || type.toUpperCase() }],
      },
      {
        type: 'paragraph',
        data: {
          hName: 'div',
          hProperties: { class: 'rspress-directive-content' },
        },
        children: children as PhrasingContent[],
      },
    ],
  };
};

/**
 * How the transformer works:
 * 1. We get the paragraph and check if it is a container directive
 * 2. If it is, crawl the next nodes, if there is a paragraph node, we need to check if it is the end of the container directive. If not, we need to push it to the children of the container directive node.
 * 3. If we find the end of the container directive, we remove the visited node and insert the custom container directive node.
 */
function transformer(tree: Root) {
  let i = 0;
  try {
    while (i < tree.children.length) {
      const node = tree.children[i];
      if (node.type !== 'paragraph') {
        i++;
        continue;
      }
      // 1. We get the paragraph and check if it is a container directive
      if (node.children[0].type !== 'text') {
        i++;
        continue;
      }
      const firstTextNode = node.children[0];
      const text = firstTextNode.value;
      const metaText = text.split('\n')[0];
      const content = text.slice(metaText.length);
      const match = metaText.match(REGEX_BEGIN);
      if (!match) {
        i++;
        continue;
      }
      const [, type, rawTitle] = match;
      // In .md, we can get :::tip{title="foo"} in the first text node
      // In .mdx, we get :::tip in first node and {title="foo"} in second node
      let title = parseTitle(rawTitle);
      // :::tip{title="foo"}
      const titleExpressionNode =
        // @ts-expect-error mdxTextExpression is not defined in mdast
        node.children[1] && node.children[1].type === 'mdxTextExpression'
          ? node.children[1]
          : null;
      // Handle the case of `::: tip {title="foo"}`
      if (titleExpressionNode) {
        title = parseTitle((titleExpressionNode as Literal).value, true);
        // {title="foo"} is not a part of the content, So we need to remove it
        node.children.splice(1, 1);
      }
      if (!DIRECTIVE_TYPES.includes(type as DirectiveType)) {
        i++;
        continue;
      }
      // 2. If it is, we remove the paragraph and create a container directive
      const wrappedChildren: BlockContent[] = [];
      // 2.1 case: with no newline between `:::` and `:::`, for example
      // ::: tip
      // This is a tip
      // :::
      // Here the content is `::: tip\nThis is a tip\n:::`
      if (content?.endsWith(':::')) {
        wrappedChildren.push({
          type: 'paragraph',
          children: [
            {
              type: 'text',
              value: content.replace(REGEX_END, ''),
            },
          ],
        });
        const newChild = createContainer(type, title, wrappedChildren);
        tree.children.splice(i, 1, newChild as Content);
      } else {
        // 2.2 case: with newline before the end of container, for example:
        // ::: tip
        // This is a tip
        //
        // :::
        // Here the content is `::: tip\nThis is a tip`
        const paragraphChild: Paragraph = {
          type: 'paragraph',
          children: [] as PhrasingContent[],
        };
        wrappedChildren.push(paragraphChild);
        if (content.length) {
          paragraphChild.children.push({
            type: 'text',
            value: content,
          });
        }
        paragraphChild.children.push(...node.children.slice(1, -1));
        // If the inserted paragraph is empty, we remove it
        if (paragraphChild.children.length === 0) {
          wrappedChildren.pop();
        }
        const lastChildInNode = node.children[node.children.length - 1];
        // We find the end of the container directive in current paragraph
        if (
          lastChildInNode.type === 'text' &&
          REGEX_END.test(lastChildInNode.value)
        ) {
          const lastChildInNodeText = lastChildInNode.value;
          const matchedEndContent = lastChildInNodeText.slice(0, -3).trim();
          // eslint-disable-next-line max-depth
          if (wrappedChildren.length) {
            (wrappedChildren[0] as Paragraph).children.push({
              type: 'text',
              value: matchedEndContent,
            });
          } else if (matchedEndContent) {
            wrappedChildren.push({
              type: 'paragraph',
              children: [
                {
                  type: 'text',
                  value: matchedEndContent,
                },
              ],
            });
          }
          const newChild = createContainer(type, title, wrappedChildren);
          tree.children.splice(i, 1, newChild as Content);
          i++;
          continue;
        }

        if (lastChildInNode !== firstTextNode && wrappedChildren.length) {
          // We don't find the end of the container directive in current paragraph
          (wrappedChildren[0] as Paragraph).children.push(lastChildInNode);
        }

        // 2.3 The final case: has newline after the start of container, for example:
        // ::: tip
        //
        // This is a tip
        // :::

        // All of the above cases need to crawl the children of the container directive node.
        // In other word, We look for the next paragraph nodes and collect all the content until we find the end of the container directive
        let j = i + 1;
        while (j < tree.children.length) {
          const currentParagraph = tree.children[j];
          if (currentParagraph.type !== 'paragraph') {
            wrappedChildren.push(currentParagraph as BlockContent);
            j++;
            continue;
          }
          const lastChild =
            currentParagraph.children[currentParagraph.children.length - 1];
          // The whole paragrph doesn't arrive at the end of the container directive, we collect the whole paragraph
          if (
            lastChild !== firstTextNode &&
            (lastChild.type !== 'text' || !REGEX_END.test(lastChild.value))
          ) {
            wrappedChildren.push({
              ...currentParagraph,
              children: currentParagraph.children.filter(
                child => child !== firstTextNode,
              ),
            });
            j++;
          } else {
            // 3. We find the end of the container directive
            // Then create the container directive, and remove the original paragraphs
            // Finally, we insert the new container directive and break the loop
            const lastChildText = lastChild.value;
            const matchedEndContent = lastChildText.slice(0, -3).trim();
            wrappedChildren.push(
              ...(currentParagraph.children.filter(
                child => child !== firstTextNode && child !== lastChild,
              ) as BlockContent[]),
            );

            if (matchedEndContent) {
              wrappedChildren.push({
                type: 'paragraph',
                children: [
                  {
                    type: 'text',
                    value: matchedEndContent,
                  },
                ],
              });
            }
            const newChild = createContainer(type, title, wrappedChildren);
            tree.children.splice(i, j - i + 1, newChild as Content);
            break;
          }
        }
      }
      i++;
    }
  } catch (e) {
    console.log(e);
    throw e;
  }
}

export const remarkPluginContainer: Plugin<[], Root> = () => {
  return transformer;
};

export default remarkPluginContainer;
