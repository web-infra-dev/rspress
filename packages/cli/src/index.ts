import { createRequire } from 'node:module';
import path from 'node:path';
import { cac } from 'cac';
import { build, dev, serve } from '@rspress/core';
import { logger } from '@rspress/shared/logger';
import chokidar from 'chokidar';
import chalk from 'chalk';
import { loadConfigFile } from './config/loadConfigFile';
import update from './update';

const CONFIG_FILES = [
  'rspress.config.ts',
  'rspress.config.js',
  '_meta.json',
  'i18n.json',
];

const require = createRequire(import.meta.url);

// eslint-disable-next-line import/no-commonjs
const packageJson = require('../package.json');

const cli = cac('rspress').version(packageJson.version).help();

const landingMessage = `ðŸ”¥ Rspress v${packageJson.version}\n`;
logger.greet(landingMessage);

const setNodeEnv = (env: 'development' | 'production') => {
  process.env.NODE_ENV = env;
};

cli.option('-c,--config [config]', 'Specify the path to the config file');

cli
  .command('[root]', 'start dev server') // default command
  .alias('dev')
  .option('--port [port]', 'port number')
  .option('--host [host]', 'hostname')
  .action(
    async (
      root,
      options?: { port?: number; host?: string; config?: string },
    ) => {
      setNodeEnv('development');
      let isRestarting = false;
      const cwd = process.cwd();
      let docDirectory: string;
      let cliWatcher: chokidar.FSWatcher;
      let devServer: Awaited<ReturnType<typeof dev>>;
      const startDevServer = async () => {
        const { port, host } = options || {};
        const config = await loadConfigFile(options?.config);

        if (root) {
          // Support root in command, override config file
          config.root = path.join(cwd, root);
        } else if (config.root && !path.isAbsolute(config.root)) {
          // Support root relative to cwd
          config.root = path.join(cwd, config.root);
        }

        docDirectory = config.root || path.join(cwd, root ?? 'docs');
        devServer = await dev({
          appDirectory: cwd,
          docDirectory,
          config,
          extraBuilderConfig: { server: { port, host } },
        });
        cliWatcher = chokidar.watch(
          [`${cwd}/**/{${CONFIG_FILES.join(',')}}`, docDirectory!],
          {
            ignoreInitial: true,
            ignored: ['**/node_modules/**', '**/.git/**', '**/.DS_Store/**'],
          },
        );
        cliWatcher.on('all', async (eventName, filepath) => {
          if (
            eventName === 'add' ||
            eventName === 'unlink' ||
            (eventName === 'change' &&
              CONFIG_FILES.includes(path.basename(filepath)))
          ) {
            if (isRestarting) {
              return;
            }
            isRestarting = true;
            console.log(
              `\nâœ¨ ${eventName} ${chalk.green(
                path.relative(cwd, filepath),
              )}, dev server will restart...\n`,
            );
            await devServer.close();
            await cliWatcher.close();
            await startDevServer();
            isRestarting = false;
          }
        });
      };

      await startDevServer();

      const exitProcess = async () => {
        await cliWatcher.close();
        await devServer.close();
      };
      process.on('SIGINT', exitProcess);
      process.on('SIGTERM', exitProcess);
    },
  );

cli.command('build [root]').action(async (root, options) => {
  setNodeEnv('production');
  const cwd = process.cwd();
  const config = await loadConfigFile(options.config);
  if (root) {
    config.root = path.join(cwd, root);
  }
  await build({
    appDirectory: cwd,
    docDirectory: config.root || path.join(cwd, root ?? 'docs'),
    config,
  });
});

cli
  .command('preview')
  .alias('serve')
  .option('--port [port]', 'port number')
  .option('--host [host]', 'hostname')
  .action(
    async (options?: { port?: number; host?: string; config?: string }) => {
      setNodeEnv('production');
      const { port, host } = options || {};
      const config = await loadConfigFile(options?.config);

      await serve({
        config,
        host,
        port,
      });
    },
  );

cli.command('update', 'update relevant packages about rspress').action(update);

cli.parse();
