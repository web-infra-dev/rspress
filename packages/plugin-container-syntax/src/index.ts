import path from 'node:path';
import { fileURLToPath } from 'node:url';
import { normalizePosixPath, type RspressPlugin } from '@rspress/shared';
import { remarkPluginContainer } from './remarkPlugin';

const __dirname = normalizePosixPath(
  path.dirname(fileURLToPath(import.meta.url)),
);

export function pluginContainerSyntax(): RspressPlugin {
  return {
    name: '@rspress/plugin-container-syntax',
    markdown: {
      remarkPlugins: [remarkPluginContainer],
    },
    globalStyles: path.posix.join(__dirname, '../container.css'),
  };
}
