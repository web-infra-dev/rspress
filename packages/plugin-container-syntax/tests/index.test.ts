import { unified } from 'unified';
import { describe, test, expect } from 'vitest';
import remarkParse from 'remark-parse';
import remarkRehype from 'remark-rehype';
import rehypeStringify from 'rehype-stringify';
import { remarkPluginContainer } from '../src/remarkPlugin';

describe('remark-container', () => {
  const processor = unified()
    .use(remarkParse)
    .use(remarkPluginContainer)
    .use(remarkRehype)
    .use(rehypeStringify);
  test('No newline', () => {
    const result = processor.processSync(`
  :::tip
  This is a tip
  :::

  12312
  `);
    expect(result.value).toMatchSnapshot();
  });

  test('With a new line after the start position of container', () => {
    const result = processor.processSync(`
  :::tip

  This is a tip
  :::`);
    expect(result.value).toMatchSnapshot();
  });

  test('With new line before the end position of container', () => {
    const result = processor.processSync(`
  :::tip
  This is a tip

  :::`);
    expect(result.value).toMatchSnapshot();
  });

  test('With new line in all case', () => {
    const result = processor.processSync(`
  :::tip

  This is a tip

  :::`);
    expect(result.value).toMatchSnapshot();
  });

  test('With code block in container', () => {
    const result = processor.processSync(`
  :::tip

  This is a tip with \`code\` and [link](foo) some text

  \`\`\`js
  const a = 1;
  \`\`\`

  :::`);
    expect(result.value).toMatchSnapshot();
  });

  test('With block quote in container', () => {
    const result = processor.processSync(`
  :::tip

  This is a tip with \`code\` and [link](foo) some text

  > This is a quote

  :::`);
    expect(result.value).toMatchSnapshot();
  });

  test('With link, inlineCode, img, list in container', () => {
    const result = processor.processSync(`
  :::tip

  This is a tip with \`code\` and [link](foo) some text

  - list 1
  - list 2

  ![img](foo)

  :::`);
    expect(result.value).toMatchSnapshot();
  });

  test('Has space before type', () => {
    const result = processor.processSync(`
  ::: tip

  This is a tip.

  :::`);
    expect(result.value).toMatchSnapshot();
  });

  test('Use {title="foo"} as title', () => {
    const result = processor.processSync(`
  ::: tip{title="Custom title"}

  This is a tip.

  :::`);
    expect(result.value).toMatchSnapshot();
  });

  test("Use {title='foo'} as title", () => {
    const result = processor.processSync(`
::: tip{title='Custom title'}

This is a tip.

:::`);
    expect(result.value).toMatchSnapshot();
  });

  test('Use {title=foo} as title', () => {
    const result = processor.processSync(`
::: tip{title=Custom title}

This is a tip.

:::`);
    expect(result.value).toMatchSnapshot();
  });

  test('Has space before :::', () => {
    const result = processor.processSync(`
::: tip title
sss
  :::`);
    expect(result.value).toMatchSnapshot();
  });

  test('details', () => {
    const result = processor.processSync(`::: details
This is a details block.
:::`);

    expect(result.value).toMatchSnapshot();
  });
});
