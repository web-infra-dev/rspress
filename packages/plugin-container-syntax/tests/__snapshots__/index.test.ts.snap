// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`remark-container > Has space before ::: 1`] = `
"<div class=\\"rspress-directive tip\\"><div class=\\"rspress-directive-title\\">title</div><div class=\\"rspress-directive-content\\"><p>
sss</p></div></div>"
`;

exports[`remark-container > Has space before type 1`] = `"<div class=\\"rspress-directive tip\\"><div class=\\"rspress-directive-title\\">TIP</div><div class=\\"rspress-directive-content\\"><p>This is a tip.</p></div></div>"`;

exports[`remark-container > No newline 1`] = `
"<div class=\\"rspress-directive tip\\"><div class=\\"rspress-directive-title\\">TIP</div><div class=\\"rspress-directive-content\\"><p>
This is a tip</p></div></div>
<p>12312</p>"
`;

exports[`remark-container > Use {title="foo"} as title 1`] = `"<div class=\\"rspress-directive tip\\"><div class=\\"rspress-directive-title\\">Custom title</div><div class=\\"rspress-directive-content\\"><p>This is a tip.</p></div></div>"`;

exports[`remark-container > Use {title='foo'} as title 1`] = `"<div class=\\"rspress-directive tip\\"><div class=\\"rspress-directive-title\\">Custom title</div><div class=\\"rspress-directive-content\\"><p>This is a tip.</p></div></div>"`;

exports[`remark-container > Use {title=foo} as title 1`] = `"<div class=\\"rspress-directive tip\\"><div class=\\"rspress-directive-title\\">Custom title</div><div class=\\"rspress-directive-content\\"><p>This is a tip.</p></div></div>"`;

exports[`remark-container > With a new line after the start position of container 1`] = `"<div class=\\"rspress-directive tip\\"><div class=\\"rspress-directive-title\\">TIP</div><div class=\\"rspress-directive-content\\"><p>This is a tip</p></div></div>"`;

exports[`remark-container > With block quote in container 1`] = `
"<div class=\\"rspress-directive tip\\"><div class=\\"rspress-directive-title\\">TIP</div><div class=\\"rspress-directive-content\\"><p>This is a tip with <code>code</code> and <a href=\\"foo\\">link</a> some text</p><blockquote>
<p>This is a quote</p>
</blockquote></div></div>"
`;

exports[`remark-container > With code block in container 1`] = `
"<div class=\\"rspress-directive tip\\"><div class=\\"rspress-directive-title\\">TIP</div><div class=\\"rspress-directive-content\\"><p>This is a tip with <code>code</code> and <a href=\\"foo\\">link</a> some text</p><pre><code class=\\"language-js\\">const a = 1;
</code></pre></div></div>"
`;

exports[`remark-container > With link, inlineCode, img, list in container 1`] = `
"<div class=\\"rspress-directive tip\\"><div class=\\"rspress-directive-title\\">TIP</div><div class=\\"rspress-directive-content\\"><p>This is a tip with <code>code</code> and <a href=\\"foo\\">link</a> some text</p><ul>
<li>list 1</li>
<li>list 2</li>
</ul><p><img src=\\"foo\\" alt=\\"img\\"></p></div></div>"
`;

exports[`remark-container > With new line before the end position of container 1`] = `
"<div class=\\"rspress-directive tip\\"><div class=\\"rspress-directive-title\\">TIP</div><div class=\\"rspress-directive-content\\"><p>
This is a tip</p></div></div>"
`;

exports[`remark-container > With new line in all case 1`] = `"<div class=\\"rspress-directive tip\\"><div class=\\"rspress-directive-title\\">TIP</div><div class=\\"rspress-directive-content\\"><p>This is a tip</p></div></div>"`;

exports[`remark-container > details 1`] = `
"<details class=\\"rspress-directive details\\"><summary class=\\"rspress-directive-title\\">DETAILS</summary><div class=\\"rspress-directive-content\\"><p>
This is a details block.</p></div></details>"
`;
